resolver/resolver.go

```diff
@@ -407,7 +407,37 @@ func isUDP(w dns.ResponseWriter) bool {
 // permitted length of the given transmission channel.
 // See https://tools.ietf.org/html/rfc1035#section-4.2.1
 func truncate(m *dns.Msg, udp bool) *dns.Msg {
-	m.Truncated = udp && m.Len() > dns.MinMsgSize
+	bufsize := 0
+	switch udp {
+	case true:
+		if opt := m.IsEdns0(); opt != nil {
+			bufsize = int(opt.UDPSize())
+		}
+		if bufsize < dns.MinMsgSize {
+			bufsize = dns.MinMsgSize
+		}
+	case false:
+		bufsize = dns.MaxMsgSize - 1
```

Comment:
I have no idea tbh, but it was introduced in skydns by @miekg and I assume he knows what he is doing :). @miekg: Care to comment? I agree that it's a bit unexpected.